azurestack/resource_arm_route.go

```diff
@@ -0,0 +1,178 @@
+package azurestack
+
+import (
+	"fmt"
+
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2015-06-15/network"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+	"github.com/terraform-providers/terraform-provider-azurestack/azurestack/utils"
+)
+
+func resourceArmRoute() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmRouteCreateUpdate,
+		Read:   resourceArmRouteRead,
+		Update: resourceArmRouteCreateUpdate,
+		Delete: resourceArmRouteDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"resource_group_name": resourceGroupNameSchema(),
+
+			"route_table_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"address_prefix": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"next_hop_type": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					string(network.RouteNextHopTypeVirtualNetworkGateway),
+					string(network.RouteNextHopTypeVnetLocal),
+					string(network.RouteNextHopTypeInternet),
+					string(network.RouteNextHopTypeVirtualAppliance),
+					string(network.RouteNextHopTypeNone),
+				}, true),
+				DiffSuppressFunc: ignoreCaseDiffSuppressFunc,
+			},
+
+			"next_hop_in_ip_address": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceArmRouteCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).routesClient
+	ctx := meta.(*ArmClient).StopContext
+
+	name := d.Get("name").(string)
+	rtName := d.Get("route_table_name").(string)
+	resGroup := d.Get("resource_group_name").(string)
+
+	addressPrefix := d.Get("address_prefix").(string)
+	nextHopType := d.Get("next_hop_type").(string)
+
+	azureStackLockByName(rtName, routeTableResourceName)
+	defer azureStackUnlockByName(rtName, routeTableResourceName)
+
+	properties := network.RoutePropertiesFormat{
```

Comment:
This could all become

```go
route := network.Route{
		Name: &name,
		RoutePropertiesFormat: &network.RoutePropertiesFormat{
			AddressPrefix: &addressPrefix,
			NextHopType:   network.RouteNextHopType(nextHopType),
		},
	}

	if v, ok := d.GetOk("next_hop_in_ip_address"); ok {
		route.RoutePropertiesFormat.NextHopIPAddress = utils.String(v.(string))
	}
```